gh-99547: Add isjunction methods for checking if a path is a junction

csm10495 opened the pull request:
Implementation of gh-99547.

This is my first try at getting something into cpython.. so please go easy on me. Thanks!

csm10495 committed:
Add an isjunction method to ntpath and posixpath, pass it along as well in pathlib

csm10495 committed:
Add is_junction test for pathlib

The actual functionality lives in ntpath/posixpath, so just verify the correct function is called and return value is returned

csm10495 committed:
Add a skipIf not win32 for testing ntpath isjunction on non windows platforms

We can only create junctions on windows, thus only test this on win32

csm10495 committed:
Merge branch 'python:main' into add_isjunction

csm10495 committed:
Merge branch 'python:main' into add_isjunction

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

csm10495 committed:
Merge branch 'python:main' into add_isjunction

eryksun commented on Lib/test/test_ntpath.py:
```diff
@@ -856,6 +857,21 @@ def test_nt_helpers(self):
+
+                # create a junction via subprocess, since we don't currently
+                # support making them via python directly
+                subprocess.check_call('mklink /J testjunc tmpdir', shell=True)
```
`CreateJunction()` is implemented for internal tests.

```suggestion
                import _winapi
                try:
                    _winapi.CreateJunction(tmpdir, 'testjunc')
                except OSError:
                    raise unittest.SkipTest('creating the test junction failed')
```

`os_helper.temp_dir()` creates the directory in the user's temp directory. This should be on the system volume, which has to be NTFS and thus support reparse points. But if for some reason creating the junction fails, the test should just be skipped.

csm10495 committed:
Update Lib/test/test_ntpath.py

Co-authored-by: Eryk Sun <eryksun@gmail.com>

csm10495 committed:
remove un-needed import

csm10495 committed:
Merge branch 'main' into add_isjunction

eryksun commented on Lib/ntpath.py:
```diff
@@ -267,6 +267,18 @@ def islink(path):
+        st = os.lstat(path)
+    except (OSError, ValueError, AttributeError):
+        return False
+    return bool(st.st_reparse_tag & stat.IO_REPARSE_TAG_MOUNT_POINT)
```
ntpath is supported on POSIX, so it needs a fallback implementation.

```suggestion
if hasattr(os.stat_result, 'st_reparse_tag'):
    def isjunction(path):
        """Test whether a path is a junction"""
        try:
            st = os.lstat(path)
        except (OSError, ValueError, AttributeError):
            return False
        return bool(st.st_reparse_tag & stat.IO_REPARSE_TAG_MOUNT_POINT)
else:
    def isjunction(path):
        """Test whether a path is a junction"""
        return False
```

csm10495 committed:
Update Lib/ntpath.py

ntpath can be used on POSIX, even though the stat_result won't have st_reparse_tag.

Co-authored-by: Eryk Sun <eryksun@gmail.com>

csm10495 committed:
fix test... this was a string not an obj

eryksun commented on Lib/posixpath.py:
```diff
@@ -169,6 +169,15 @@ def islink(path):
+def isjunction(path):
+    """Test whether a path is a junction
+    Junctions are not a part of posix semantics"""
+    return False
```
Calling `os.lstat()` implicitly calls `os.fspath()`, which raises `TypeError` if the path argument isn't a string, bytes, or path-like. Even though it isn't needed normally, `os.fspath()` still has to be called on POSIX.
   
```suggestion
    os.fspath(path)
    return False
```

csm10495 committed:
Update Lib/posixpath.py

Co-authored-by: Eryk Sun <eryksun@gmail.com>

eryksun commented on Lib/ntpath.py:
```diff
@@ -267,6 +267,23 @@ def islink(path):
+else:
+    def isjunction(path):
+        """Test whether a path is a junction"""
+        return False
```
```suggestion
        os.fspath(path)
        return False
```

See the similar suggestion in posixpath.py for the reason.

csm10495 committed:
Update Lib/ntpath.py

Co-authored-by: Eryk Sun <eryksun@gmail.com>

csm10495 committed:
Empty commit to rekick ci

csm10495 committed:
update os.path.rst for isjunction

csm10495 committed:
update pathlib docs for is_junction

csm10495 committed:
Merge pull request #1 from csm10495/add_isjunction

Add isjunction

csm10495 committed:
Merge pull request #2 from csm10495/main

csm10495 committed:
Merge branch 'main' into add_isjunction

zooba commented on Doc/library/pathlib.rst:
```diff
@@ -889,6 +889,14 @@ call fails (for example because the path doesn't exist).
+.. method:: Path.is_junction()
+
+   Return ``True`` if the path points to a junction, and ``False`` for any other
+   type of file (as returned by :func:`os.path.isjunction`).
```
Note here that junctions currently only exist on Windows, but the function is available on all platforms. (Preferably in a way like that which allows us to expand the definition in the future if necessary).

I think we'd rather not have the reference back to os.path here. The pathlib docs should have all the useful information the user needs.

csm10495 replied:
Edit: <replied in wrong place> moving this reply up the chain.

zooba commented on Doc/library/os.path.rst:
```diff
@@ -266,6 +266,14 @@ the :mod:`glob` module.)
+.. function:: isjunction(path)
+
+   Return ``True`` if *path* is a junction. This function exists for both POSIX
+   and Windows, though can only return True on Windows.
```
```suggestion
   and Windows, though only Windows currently supports junctions.
```

This allows us to support junctions on other platforms if it ever comes up. Making the details too concrete would mean we'd have to deprecate the function before allowing another platform to return True for their junctions.

csm10495 committed:
Update 3.12.rst

csm10495 committed:
Merge pull request #3 from csm10495/csm10495-patch-1

Update 3.12.rst

zooba reviewed (commented):
Minor tweaks (since we treat docs as a specification, so it pays to be precise).

I wonder whether we should be using "directory junctions" rather than just "junctions"? I'm not sure if there are any other types, but the test would be "if there's another type of junction, would `isjunction` still return true?"

As implemented, it would not, so always referring to "directory junctions" would be fine. But if there's a possibility of stretching the definition in the future, we can be a bit more general and just call them "junctions".

csm10495 committed:
Update Doc/library/os.path.rst

Co-authored-by: Steve Dower <steve.dower@microsoft.com>

csm10495 committed:
Update pathlib.rst

csm10495 committed:
Update pathlib.rst

csm10495 committed:
Update os.path.rst

csm10495 committed:
Update 3.12.rst

csm10495 committed:
Update os.path.rst

eryksun commented on Lib/ntpath.py:
```diff
@@ -267,6 +267,24 @@ def islink(path):
+            st = os.lstat(path)
+        except (OSError, ValueError, AttributeError):
+            return False
+        return bool(st.st_reparse_tag & stat.IO_REPARSE_TAG_MOUNT_POINT)
```
Sorry, using a bitwise AND here was an editing mistake in my original message on python-ideas. This needs to be an equality check:

```suggestion
        return bool(st.st_reparse_tag == stat.IO_REPARSE_TAG_MOUNT_POINT)
```
---

The upper 16 bits of a reparse tag are flags, of which 12 bits are reserved, and the lower 16 bits are the assigned tag number in one of two domains: Microsoft and non-Microsoft. For example, `IO_REPARSE_TAG_MOUNT_POINT` (0xA000_0003) and `IO_REPARSE_TAG_SYMLINK` (0xA000_000C) have the flags set to indicate a Microsoft tag (0x8000_0000) and name-surrogate type (0x2000_0000). Currently there are about 50 Microsoft tags and about 80 non-Microsoft tags.

csm10495 committed:
Update Lib/ntpath.py

Co-authored-by: Eryk Sun <eryksun@gmail.com>

csm10495 committed:
pr fixes, also add contribution notices

csm10495 committed:
Swap back to LF line endings

csm10495 committed:
Merge branch 'main' into add_isjunction

csm10495 committed:
update docs

csm10495 committed:
fix mistype

csm10495 committed:
Add test for DirEntry is_junction()

csm10495 committed:
rekick ci

zooba reviewed (approved):
I'm happy with this, as soon as you say that you're done. Great work!

eryksun commented on Doc/library/os.rst:
```diff
@@ -2738,6 +2738,17 @@ features:
+
+      Return ``True`` if this entry is a junction (even if broken);
+      return ``False`` if the entry points to a directory or any kind of file,
+      or if it doesn't exist anymore.
```
```suggestion
      return ``False`` if the entry points to a regular directory or file, a
      symlink, or if it doesn't exist anymore.
```

Unlike directory symlinks, junctions are reported as directories. They're mount points, after all.

Note however that junctions can be deleted like files via `os.unlink()`, for which we call WinAPI `RemoveDirectoryW()`. I think that's POSIX-ly correct because the spec for `unlink()` says "[t]he path argument shall not name a directory unless the process has appropriate privileges and the implementation supports using unlink() on directories". One could argue that an implicit privilege enables `unlink()` to delete junctions.

csm10495 committed:
Update Doc/library/os.rst

Co-authored-by: Eryk Sun <eryksun@gmail.com>

eryksun commented on Doc/library/os.rst:
```diff
@@ -2738,6 +2738,17 @@ features:
+
+      Return ``True`` if this entry is a junction (even if broken);
+      return ``False`` if the entry points to a regular directory or file, a
+      symlink, or if it doesn't exist anymore.
```
Oops, I dropped the "any kind" distinction for files.

```suggestion
      return ``False`` if the entry points to a regular directory, any kind
      of file, a symlink, or if it doesn't exist anymore.
```

csm10495 committed:
Update Doc/library/os.rst

Co-authored-by: Eryk Sun <eryksun@gmail.com>

eryksun commented on Lib/test/test_os.py:
```diff
@@ -4206,6 +4208,21 @@ def test_attributes(self):
+
+        import _winapi
+        try:
+            _winapi.CreateJunction(os.path.join(self.path, "srcjunc"), dirname)
```
```suggestion
            _winapi.CreateJunction(dirname, os.path.join(self.path, "srcjunc"))
```

`CreateJunction()` uses the same parameter order as `os.symlink()`. The first parameter is the target path. The second parameter is the path of the junction to be created. Passing the existing directory `dirname` as the second argument causes the call to raise `FileExistsError`, and the test gets skipped.

csm10495 committed:
Update Lib/test/test_os.py

Co-authored-by: Eryk Sun <eryksun@gmail.com>

csm10495 committed:
kick ci

csm10495 committed:
Remove _rmtree_is_dir() since we can now just check if the entry is a dir and not a junction across the board

eryksun commented on Lib/shutil.py:
```diff
@@ -605,7 +587,7 @@ def _rmtree_unsafe(path, onerror):
     for entry in entries:
         fullname = entry.path
-        if _rmtree_isdir(entry):
+        if entry.is_dir(follow_symlinks=False) and not entry.is_junction():
```
```suggestion
        try:
            is_dir = (entry.is_dir(follow_symlinks=False) and
                      not entry.is_junction())
        except OSError:
            is_dir = False
```

The `is_dir()` method call might raise `OSError` on a POSIX platform that lacks the `d_type` field in a directory entry[^1] (e.g. IBM AIX) -- or if the value is `DT_UNKNOWN` because the filesystem doesn't support the `d_type` field. In this case `stat()` has to be called, for which the internal `DirEntry_test_mode()` function only ignores `FileNotFoundError`. We can assume it's not a directory if any `OSError` is raised, just as `os.path.isdir()` does.

[^1]: POSIX doesn't specify `d_type` in a [`dirent` record](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/dirent.h.html). It's found on BSD systems and Linux.

csm10495 committed:
Update Lib/shutil.py

Co-authored-by: Eryk Sun <eryksun@gmail.com>

csm10495 committed:
update from suggestion

eryksun commented on Lib/shutil.py:
```diff
@@ -605,7 +587,13 @@ def _rmtree_unsafe(path, onerror):
+        except OSError:
+            is_dir = False
+
+        if is_dir:
```
Since the `is_junction()` method call can't raise `OSError`, you could move that check out of the try/except. I think this is less muddled since a junction is a directory.

```suggestion
        try:
            is_dir = entry.is_dir(follow_symlinks=False)
        except OSError:
            is_dir = False

        if is_dir and not entry.is_junction():
```

csm10495 committed:
Update Lib/shutil.py

Co-authored-by: Eryk Sun <eryksun@gmail.com>
